### PR TITLE
Rewrite embulk-input-jira in Java

### PR DESCRIPTION
**Pull request by minidragon88**
Java implementation for embulk-input-jira

### CODE REVIEW

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Code style: should we clean these left-over comments up?

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Code style: same as above

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: Code style: would be simpler to write like this instead of ternary:
```
arrayObj.size() > 0 &&
StreamSupport.stream(arrayObj.spliterator(), false).allMatch(x -> x.isJsonObject());
```
The below are purely personal taste, I just think these are nicer:
`x -> x.isJsonObject()` could be replaced by a method reference `JsonElement::isJsonObject`;
StreamSupport.stream can be import statically, since `stream()` method is already obvious on what it does.

Comment (huylenq) [reply src/main/java/org/embulk/input/jira/Issue.java]: Question: the reason we have to explicitly make this a sequential stream?

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: Same question about parallel as above. In this case is it because `JsonArray` can not take concurrent writes?

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java:70]: This method definitely need  to be unit tested

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: Just using plus `+` operators would be easier to read (and actually perform a little bit better). Also, inline this `prefix.isEmpty() ? key : prefix + "." + key` to the caller would save a method dispatch.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/Issue.java]: Code style: IMO,
```
for (JsonElement el : arrayObj) {
    newObj.get(key).getAsJsonArray().add(el.getAsJsonObject().get(key));
}
```
Is more ease to the eyes.

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/client/JiraClient.java]: Since this is a user facing message: "due to problem" should be "due to a problem" or "due to problems".

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/client/JiraClient.java:229]: Should we be defensive here? Expect a `AuthenticationMethod.BASIC` case, and throw otherwise. So adding additional authentication methods later won't be a behavior change (avoid the case where somehow there is an invalid `auth_method` worked in the past but won't work after upgraded).

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/client/JiraClient.java]: A little hairy and seems like incorrect logic, unauthorized is retry-able?
Put correctness aside, this could be simplified / translated as-is to:
```
return statusCode / 100 != 4 || statusCode == HttpStatus.SC_UNAUTHORIZED || statusCode == 429;
```

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/client/JiraClient.java]: Code style: just wording, "Completed" sounds a little bit misleading, imo "Retry Limit Exceeded" would be easier to understand from the logs.

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/client/JiraClient.java]: Code style: should remove the redundant `client` parameter?

**Review by huylenq — CHANGES_REQUESTED** [review]
All the comments that starts with "Code style:" are only my opinions to consider.
The rest probably need to be addressed before merging.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: Code style: naming `fetch` is quite misleading, I thought it involving IO. Using `get` here could be more common (lodash exactly using `get` to lookup for value by dot separated path)

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: Code style: already mentioned, prefering streams when they could be composed. In this case, where it is for side effect and `JsonArray` is already a `Iterable`, I think a simple `for` loop would be easier to read. If we really want to leverage Java Stream then:
```
JsonArray arrays = StreamSupport.stream(json.getAsJsonArray().spliterator(), false)
        .map(obj -> obj.isJsonObject() ? obj.getAsJsonObject().get(key) : obj)
        .collect(Collector.of(
                JsonArray::new, JsonArray::add,
                (JsonArray acc1, JsonArray acc2) -> { acc1.addAll(acc2); return acc1; }));
```
is more idiomatic I guess.

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/util/JiraUtil.java]: Code style: could be simplified with `Collectors#joining`

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Code style: `List` already have `forEach`, lambda could be replace by method reference: `Issue::toRecord`.

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Could be more stream-ish with:
```
return issues.stream()
        .flatMap(issue -> issue.getFlatten().keySet().stream())
        .collect(Collectors.toSet());   // or better readability with static import, then: .collect(toSet())
```
That would return a `Set` in interface and `HashSet` implementation, not `SortedSet`. But look like the we don't rely on the order of the result anyway?

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Code style: Again, `List` already has `forEach`, if we really want to use streams, then this is a little more idiomatic:
```
return issues.stream()
        .map(Issue::getFlatten)
        .map(flattened -> {
            JsonObject unified = new JsonObject();
            for (String attr : uniqAtrribtes) {
                unified.add(attr, flattened.has(attr) ? flattened.get(attr) : JsonNull.INSTANCE);
            }
            return unified;
        })
        .collect(Collector.of(
                JsonArray::new, JsonArray::add,
                (acc1, acc2) -> { acc1.addAll(acc2); return acc1; }));
```
We generally should avoid mixing streams with side-effects if possible. (you also had a spelling error there: `uniqAttribtes`, I retained it on the example above)

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Done. 
The reason I do not use parallel because the size normally small.
Likely it is < 10. Sometimes only 1 or 2.
Hence using parallel is a waste. I suppose.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Done.
The question is answer in above.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java:70]: I have added unit tests.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Done

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: Done

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java:229]: I think we could prevent this inside of AuthenticateMethod enum. 
And I've already has it in https://github.com/treasure-data/embulk-input-jira/blob/embulk-input-jira_java/src/main/java/org/embulk/input/jira/AuthenticateMethod.java#L24

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: Done.
The reason when we retry on 401 is because these could be 401 status when the user try to overloading JIRA server. Which we have to do the fix https://github.com/treasure-data/embulk-input-jira/pull/52 for ruby version.

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: Done

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: Done

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Change to get and getValue

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Done

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/util/JiraUtil.java]: Collectors#joining is for string.
When I changed the test failed. Hence I would like to keep it as current

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Changed to for in.
The reasons I need sortedSet is to ensure the property will well ordered when we return the config diff to user.

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Changed to for in

**Comment by minidragon88** [comment]
@huylenq could you take a look again on the PR?

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/util/JiraUtil.java]: This failed?
```
pageBuilder.setString(column,
        StreamSupport
                .stream(data.getAsJsonArray().spliterator(), false)
                .map(obj -> {
                    if (obj.isJsonPrimitive()) {
                        return obj.getAsString();
                    }
                    else {
                        return obj.toString();
                    }
                })
                .collect(Collectors.joining(",")));
```

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java:229]: Got it, but I still don't understand why if the logic of this part of `JiraClient` is if the `authMethod` is `BASIC` then append the "Authorization: Basic ..." header then why isn't it simply written so but "catch all" with `default`? It might make readers puzzled with "Is the intention is to fallback to Basic authorization if the known authorization is unhandled/unimplemented?"

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: Understood, but leaving a comment to explain that would be very helpful. The retrying status code range is a bit puzzled to me, retrying all except 4xx (surely with the exception of 401 and 429), but I'm don't know enough the JIRA specs to tell.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/Issue.java]: These mutable, stateful methods usually prompt for errors. Is there anything preventing us from computing `flatten` when `getFlatten()` is called (and certainly cache it), instead of implicitly relying on that `toRecord` must had been called earlier?

**Review by huylenq — CHANGES_REQUESTED** [review]

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Nitpick: "Atrribtes", there are several places having the same spelling mistake.

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java:148]: Looks like we inherit a very small portion of "embulk-parser-jsonpath" (which actually just delegate to embulk's built-in's `SchemaGuess`) for the guess feature. Writing our own ruby function to leverage `SchemaGuess` seems like a leaner way comparing to this. Our own version, arguably, would be more simple, we know exactly where is the JSON root, whether it's an array or object and how the guess result diff should be shaped.

This is just a suggestion, but please consider, I think could save later maintainers some time for they need to figure out why an input plugin directly requires a parser plugin.

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/Issue.java]: Should we so defensive here ?
In the beginning of my development, I put it in the constructor. 
But after that I realize this method only trigger on guessing. And I can freely control whenever to call it. So I decided to put it as it is.

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java:148]: Sincerely I would enjoy the code from the other plugin.
Write it is not a problem but we could potential duplicate code from both https://github.com/treasure-data/embulk-input-jira/blob/master/lib/embulk/input/jira_input_plugin_utils.rb#L30 and https://github.com/embulk/embulk/blob/57b42c31d1d539177e1e818f294550cde5b69e1f/lib/embulk/guess/schema_guess.rb#L16-L24%20def%20from_hash_records(array_of_hash)%20array_of_hash%20=%20Array(array_of_hash)%20if%20array_of_hash.empty?%20raise%20%22SchemaGuess%20Can't%20guess%20schema%20from%20no%20records%22%20end%20column_names%20=%20array_of_hash.map(&:keys).inject([])%20{|r,%20a|%20r%20+%20a%20}.uniq.sort%20samples%20=%20array_of_hash.to_a.map%20{|hash|%20column_names.map%20{|name|%20hash[name]%20}%20}%20from_array_records(column_names,%20samples)%20end%20end%20end%20end%20module%20Embulk%20module%20Input%20module%20JiraInputPluginUtils%20#%20Guess::SchemaGuess.from_hash_records%20returns%20Columns%20#%20containing%20'index'%20key,%20but%20it%20is%20needless.%20def%20self.guess_columns(records)%20schema%20=%20Guess::SchemaGuess.from_hash_records(records)%20schema.map%20do%20|c|%20column%20=%20{name:%20c.name,%20type:%20c.type}%20column[:format]%20=%20c.format%20if%20c.format%20column%20end%20end%20def%20self.cast(value,%20type)%20return%20value%20if%20value.nil?%20case%20type.to_sym%20when%20:long%20Integer(value)%20when%20:double%20Float(value)%20when%20:timestamp%20Time.parse(value)%20when%20:boolean%20!!value%20else%20value.to_s%20end%20end%20end%20end%20end
So I decide to use the existing one.
As https://github.com/embulk/embulk/blob/v0.9.12/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java#L73-L86 also do the same

File: src/main/java/org/embulk/input/jira/Issue.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/Issue.java]: It's just that the current implementation has room for errors, while the alternative one with pretty much same number of lines of code doesn't. But it doesn't mean your current one is incorrect and the decision is yours, I consider this is resolved 👍

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java:148]: Your call. But I don't think we're on the same page, let's take the discussion offline.

**Review by huylenq — APPROVED** [review]

**Comment by huylenq** [comment]
LGTM

File: embulk-input-jira.gemspec
Comment (huylenq) [inline embulk-input-jira.gemspec]: One more thing, should we remove this file from Git?

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: From @tvhung83 (extracted from https://github.com/treasure-data/embulk-input-jira/pull/54#discussion_r257113546):
> Not related to this PR, but please make it a habit by not introducing a variable when you don't really need it. For example, `#createGuessConfig` and `#getGuessExecutor` could be expensive, what a waste if `#validateTaskConfig` fails.

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: From @tvhung83 (extracted from https://github.com/treasure-data/embulk-input-jira/pull/54#discussion_r257113913)
```suggestion
         return Exec.newConfigSource()
            .set("guess_plugins", ImmutableList.of("jira"))
            .set("guess_sample_buffer_bytes", GUESS_BUFFER_SIZE);
```

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/JiraInputPlugin.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/JiraInputPlugin.java]: Done

File: src/main/java/org/embulk/input/jira/client/JiraClient.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/client/JiraClient.java]: The reason why we retry on 401 is because when we try to have so many requests in once to JIRA APIs (i.e 100 requests). Then the JIRA APIs will return 401 error although the credential is correct. 
Hence we need to retry on such situation.
Otherwise 429 is stand for Too many request. We also need to retry.
Other 4xx error are considered errors.

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/util/JiraUtil.java]: Done

**Comment by huylenq** [comment]
Recent changes up to fd2f1ee also LGTM

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (huylenq) [inline src/main/java/org/embulk/input/jira/util/JiraUtil.java]: Nitpick: the trailing dot should be on the next line. Purely my taste, not necessary to change: the builder style for multiple segments benefits when it is mixed with variables; for constants, `...fromUri(url).path("rest/api/latest/myself")...` looks a bit more concise and still easy to read.

**Review by huylenq — APPROVED** [review]
https://github.com/treasure-data/embulk-input-jira/compare/fd2f1ee..8ba81d0 LGTM.

File: src/main/java/org/embulk/input/jira/util/JiraUtil.java
Comment (minidragon88) [reply src/main/java/org/embulk/input/jira/util/JiraUtil.java]: Done

**Comment by huylenq** [comment]
https://github.com/treasure-data/embulk-input-jira/compare/8ba81d0..f28889a LGTM

**Review by huylenq — APPROVED** [review]

**Pull request merged**
